File: code/73.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-07-05 19:21:50
LastEditors: lmio 2091319361@qq.com
Description: 73. 矩阵置零
'''

from typing import List

def setZeroes(matrix: List[List[int]]) -> None:
    row = set()
    col = set()
    for i in range(len(matrix)):
        for j in range(len(matrix[i])):
            if matrix[i][j] == 0:
                row.add(i)
                col.add(j)
    for v in row:
        for j in range(len(matrix[v])):
            matrix[v][j] = 0
    for v in col:
        for i in range(len(matrix)):
            matrix[i][v] = 0

File: code_cpp/73.cpp
#include <bits/stdc++.h>

using namespace std;

struct Solution {
    void setZeroes(vector<vector<int>> &matrix) {
        bool col0_flag = false;
        int row = matrix.size();
        int col = matrix[0].size();
        for (int i = 0; i < row; i++) {
            if (matrix[i][0] == 0)
                col0_flag = true;
            for (int j = 1; j < col; j++) {
                if (matrix[i][j] == 0) {
                    matrix[i][0] = matrix[0][j] = 0;
                }
            }
        }
        for (int i = row - 1; i >= 0; i--) {
            for (int j = col - 1; j >= 1; j--) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
            if (col0_flag)
                matrix[i][0] = 0;
        }
    }
};